File: TerraObserver/TO.Abstractions/Views/Cameras/IOrbitCameraRig.cs
using Godot;
using Godot.Abstractions.Bases;

namespace TO.Abstractions.Views.Cameras;

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-03 17:38:03
public interface IOrbitCameraRig : INode3D
{
    #region 事件

    event Action<float>? Processed;
    event Action? ZoomChanged;

    delegate void MovedEvent(Vector3 pos, float delta);

    event MovedEvent? Moved;
    void EmitMoved(Vector3 pos, float delta);

    delegate void TransformedEvent(Transform3D transform, float delta);

    event TransformedEvent? Transformed;
    void EmitTransformed(Transform3D trans, float delta);

    #endregion

    #region Export

    float StickMinZoom { get; }
    float StickMaxZoom { get; }
    float SwivelMinZoom { get; }
    float SwivelMaxZoom { get; }
    float MoveSpeedMinZoom { get; }
    float MoveSpeedMaxZoom { get; }
    float RotationSpeed { get; }
    Node3D? Sun { get; }
    float AutoPilotSpeed { get; }

    #endregion

    #region on-ready

    Node3D FocusBase { get; }
    CsgBox3D FocusBox { get; }
    Node3D FocusBackStick { get; }
    CsgBox3D BackBox { get; }
    SpotLight3D Light { get; }
    Node3D Swivel { get; }
    Node3D Stick { get; }
    RemoteTransform3D CamRig { get; }

    #endregion

    #region 普通属性

    float Zoom { get; set; }
    float AntiStuckSpeedMultiplier { get; set; }
    float AutoPilotProgress { get; set; }
    Vector3 FromDirection { get; set; }
    Vector3 DestinationDirection { get; set; }

    #endregion
}

File: TerraObserver/TO.Domains/Types/Cameras/IOrbitCameraRig.fs
namespace TO.Domains.Types.Cameras

open Godot
open TO.Domains.Types.Godots

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-29 13:22:29
[<Interface>]
type IOrbitCameraRig =
    inherit INode3D
    // =====【事件】=====
    abstract EmitMoved: Vector3 -> float32 -> unit
    abstract EmitTransformed: Transform3D -> float32 -> unit
    // =====【Export】=====
    abstract StickMinZoom: float32
    abstract StickMaxZoom: float32
    abstract SwivelMinZoom: float32
    abstract SwivelMaxZoom: float32
    abstract MoveSpeedMinZoom: float32
    abstract MoveSpeedMaxZoom: float32
    abstract RotationSpeed: float32
    abstract Sun: Node3D
    abstract AutoPilotSpeed: float32
    // =====【on-ready】=====
    abstract FocusBase: Node3D
    abstract FocusBox: CsgBox3D
    abstract FocusBackStick: Node3D
    abstract BackBox: CsgBox3D
    abstract Light: SpotLight3D
    abstract Swivel: Node3D
    abstract Stick: Node3D
    abstract CamRig: RemoteTransform3D
    // =====【普通属性】=====
    abstract Zoom: float32 with get, set
    abstract AntiStuckSpeedMultiplier: float32 with get, set
    abstract AutoPilotProgress: float32 with get, set
    abstract FromDirection: Vector3 with get, set
    abstract DestinationDirection: Vector3 with get, set
